stl/inc/memory

```diff
@@ -4121,6 +4121,177 @@ public:
+    using type = typename _Ty::element_type*;
+};
+
+template <class _Ty>
+    requires (!_Has_member_element_type<_Ty> && !_Has_member_pointer<_Ty>
+        && _Has_member_element_type<pointer_traits<_Ty>>)
+struct _Pointer_of_helper<_Ty> {
+    using type = typename pointer_traits<_Ty>::element_type*;
+};
+// clang-format on
+
+template <class _Ty>
+using _Pointer_of = typename _Pointer_of_helper<_Ty>::type;
+
+template <class _Ty, class _Uty>
+struct _Pointer_of_or_helper {
+    using type = _Uty;
+};
+
+// clang-format off
+template <class _Ty, class _Uty>
+    requires requires { typename _Pointer_of<_Ty>; }
+struct _Pointer_of_or_helper<_Ty, _Uty> {
+    using type = _Pointer_of<_Ty>;
+};
+// clang-format on
+
+template <class _Ty, class _Uty>
+using _Pointer_of_or = typename _Pointer_of_or_helper<_Ty, _Uty>::type;
+
+// TRANSITION, requires expression support
+template <class _SmartPtr, class _Sp, class _Pointer, class... _ArgsT>
+concept _Resettable_pointer = requires(_SmartPtr& _Smart_ptr, _Pointer _Ptr, _ArgsT&&... _Args) {
+    _Smart_ptr.reset(static_cast<_Sp>(_Ptr), _STD forward<_ArgsT>(_Args)...);
+};
+
+template <class _SmartPtr, class _Pointer, class... _ArgsT>
+class out_ptr_t {
+    static_assert(!_Is_specialization_v<_SmartPtr, shared_ptr> || sizeof...(_ArgsT) != 0,
+        "out_ptr_t with shared_ptr requires a deleter");
+
+public:
+    explicit out_ptr_t(_SmartPtr& _Smart_ptr_, _ArgsT... _Args_) noexcept(
+        is_nothrow_constructible_v<tuple<_ArgsT...>, _ArgsT...>)
+        : _Smart_ptr(_Smart_ptr_), _Args(_STD forward<_ArgsT>(_Args_)...) {}
+
+    out_ptr_t(const out_ptr_t&) = delete;
+
+    ~out_ptr_t() {
+        if (!_Ptr) {
+            return;
+        }
+
+        _STD apply(
+            [this](auto&&... _Args_) {
+                using _Sp = _Pointer_of_or<_SmartPtr, _Pointer>;
+                if constexpr (_Resettable_pointer<_SmartPtr, _Sp, _Pointer, _ArgsT...>) {
+                    _Smart_ptr.reset(static_cast<_Sp>(_Ptr), _STD forward<_ArgsT>(_Args_)...);
+                } else {
+                    static_assert(is_constructible_v<_SmartPtr, _Sp, _ArgsT...>);
+                    _Smart_ptr = _SmartPtr(static_cast<_Sp>(_Ptr), _STD forward<_ArgsT>(_Args_)...);
+                }
+            },
+            _STD move(_Args));
+    }
+
+    operator _Pointer*() const noexcept {
+        return _STD addressof(const_cast<_Pointer&>(_Ptr));
+    }
+
+    operator void**() const noexcept requires(!is_same_v<_Pointer, void*>) {
+        static_assert(is_pointer_v<_Pointer>,
+            "conversion of out_ptr_t<Smart, Pointer, Args...> to void** requires Pointer to be a raw pointer");
+        return reinterpret_cast<void**>(_STD addressof(const_cast<_Pointer&>(_Ptr)));
+    }
+
+private:
+    _SmartPtr& _Smart_ptr;
+    tuple<_ArgsT...> _Args;
+    _Pointer _Ptr{};
+};
+
+template <class _Pointer = void, class _SmartPtr, class... _ArgsT>
+_NODISCARD auto out_ptr(_SmartPtr& _Smart_ptr, _ArgsT&&... _Args) {

```

Comment:
Should this function be strengthened, if strengthening `out_ptr_t`'s constructor was important?

(According to my understanding, this could be strengthened to unconditional `noexcept`, because it's passing lvalue/rvalue references for the args, and constructing a tuple of references is `noexcept` - that might deserve a comment as it is subtle. The `out_ptr_t` constructor strengthening has to be conditional `noexcept` in order to handle the case where the user explicitly passes object types instead of reference types.)

Reply:
This can only be conditional on `is_nothrow_constructible` - we permit user specializations of `out_ptr_t` that don't even have to meet the requirements of the primary template.